### ISSUE TITLE
Tracking Interfaces

### ISSUE THREAD

**Issue by ordinaryorange** [mfidemraizer/trackerdog#7]
The following seems natural
``` C#
     void Main()
    {
    	TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<A>());
    	var alph = new Alpha();
    	var t = alph.AsTrackable();
    	t.MyProperty = 100;
    	var tracker = t.GetChangeTracker(); //Contract Excetption not a trackable object	
    }

    public interface A
    {
    	int MyProperty { get; set; }
    }

    public class Alpha : A
    { 
    	public virtual int MyProperty { get; set; }
        public virtual bool HisProperty {get ; set;}
    }
```

but produces the noted exception.
I think that MyProperty should be tracked, but not HisProperty.
Could it even be possible for `var t = Tackable.Of<A>()` and then get the assistance from a dependency injection container to create the proxy ?

**Comment by mfidemraizer** [comment]
Uhm I believe you'll cry but this time it seems like our favorite dog isn't mistaken!

`A` is trackable but you've not configured `Alpha` as trackable, have you? :)

TrackerDog needs you to configure the entire inheritance tree. I've been thinking about something like this that could be a new feature:

`TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<Alpha>(configureInheritance: true));`

...which would simplfiy get all `Alpha`'s base types and would configure them for you. BTW, the main issue with this is that this is fantastic if you don't want to configure which properties to track.... but If you want to... this feature is useless and this is the main reason to forcing everyone to configure an inheritance type by type... (FOR NOW). I could implement this feature anyway if you just want to configure the entire graph without customizing which properties to track.

Please check if adding the following line solves your issue:

```
TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<A>());
TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<Alpha>());
```

**Comment by mfidemraizer** [comment]
About the other question...

> Could it even be possible for var t = Tackable.Of<A>() and then get the assistance from a dependency injection container to create the proxy ?

In the current form, it might depend on your dependency injection container. For example, in Castle Windsor you could use `UseFactoryMethod` to define how a component gets instantiated:

`container.Register(Component.For<Alpha>().UsingFactoryMethod(() => Trackable.Of<A>()));`

This way, you can even configure the proxy life-cycle:

`container.Register(Component.For<Alpha>().UsingFactoryMethod(() => Trackable.Of<A>()).LifestyleTransient());`

...and many other options.

AFAIK, other DI containers also support component factories.

What I'm already considering thanks to some suggestion of some friend who's already using TrackerDog like you is refactoring configuration to be DI-friendly.

Instead of working with static configurations, it would be nice to have something like this:



And this configuration would be injectable using a DI container:

```
container.Register
(
        Component.For<ITrackingFactory>().UsingFactoryMethod
        (
                () =>
                {
                           TrackingConfiguration config = new TrackingConfiguration();
                           config.TrackThisType<A>();
                           config.TrackThisType<B>();

                          return config.Build();
                }
        )
);
```

...so you can inject `ITrackingConfiguration` on your dependencies:

```
public class X
{
       public X(ITrackableObjectFactory tracking)
       {
                 A trackableA = tracking.Of<A>();
       }
}
```

BTW I'll need some time to implement this refactor. It's easy to implement but it'll take some hours ;P

**Comment by ordinaryorange** [comment]
In regard to the first bit of this thread. I intentionally did not set alpha in the tracking configuration.
Here is my thinking.
Say you have an interface such as 
``` c#
public interface IPerson
{
    bool IsHappy {get;}
}
```
I could create instances of `Child`, `Adult`, `Eskimo`, `Businessman`..etc all which implement `IPerson`.
I don't care about any of the other properties each of those Types expose, but if an object implements `IPerson` then please track the implemented properties of that interface eg.
``` c#
public class MyList : List<IPerson>
{
    public void Add(IPerson person)
    {
        base.Add(person.AsTrackable());
    }
}
```
This would really make tracker dog easy to incorporate rather than having to pre configure the concrete types.
This then leads onto the DI discussion. If a DI container knows how to "new" up a IPerson, and if we pass the container to Tracker Dogs configuration, Tracker dog can ask the container for a new instance of an object it needs to wrap a proxy around. And a call to `var t = Trackable.Of<IPerson>()` would work without the need to inject in a `ITrackableObjectFactory`.
Again all just suggestion with my youth of knowledge about the inner workings of tracker dog.

**Comment by mfidemraizer** [comment]
Oh, I understand what you mean. In summary, you want to use interfaces like *configurations*. If you call `x.AsTrackable()` where `x` is a reference to an object which implements an already configured interface, then interface's properties should be *trackable*.

I'll check the source code to see if this is possible.

**Comment by ordinaryorange** [comment]
Bingo you got it. I hope it's not too difficult to implement. The configuration via interfaces will help reduce dependencies when integrating TrackerDog.

**Comment by mfidemraizer** [comment]
Nice! Well, I'll implement it as soon as possible. I understand that you can go with configuring each type one by one in the mean time, can't you?

It shouldn't be hard to implement. Today will be impossible, but I expect to be able to implement this in the next days!

**Comment by FranckDepoortere** [comment]
Hi, 
I guess I've got the same problem than  ordinaryorange:
I'd like to register interfaces and not final Class Types: 
the reason is that I don't have access to these Classes in the place where I register configuration TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<A>());
I just have access to the interfaces.

It would be very useful to register Interfaces.
Thanks 
Franck D.

**Comment by mfidemraizer** [comment]
@FranckDepoortere Yeah, you both are right about this feature. This is the nice part of open sourcing project: since you don't know every use case that others might or mightn't need, you publish a basic feature set and it gets enhanced overtime!

So... I'm thinking about two possible ways of implementing this...

### Enhancing registration API

For example:

``
TrackerDogConfiguration.TrackTheseTypes(Track.ImplementationsOf<IWhatever>().FromAssemblies("YourProject.A", "YourProject.B"));
``

Actually this would be less effort because it's just that TrackerDog would search implementations of the whole interface in some given assemblies. Maybe it would be nice to filter found types to be able to implement some kind of convention:

``
TrackerDogConfiguration.TrackTheseTypes(Track.ImplementationsOf<IWhatever>().FromAssemblies("YourProject.A", "YourProject.B").Filter(type => type.Name.StartsWith("Domain")));
``

I could also implement:

``
// This would register types implementing IWhatever
// from current executing assembly
TrackerDogConfiguration.TrackTheseTypes(Track.ImplementationsOf<IWhatever>().FromThisAssembly());
``

### 2. Lazy registration

This approach would need some extra effort. For example, you would register an interface type:

``
TrackerDogConfiguration.TrackTheseTypes(Track.ThisInterface<IWhatever>());
``

During run-time, you would code an implementation of `IWhatever` and you would call `AsTrackable()`:

``
WhateverImpl impl = new WhateverImpl().AsTrackable();
``

When the whole implementation hits `AsTrackable()`, TrackerDog would be smart enough to check that `WhateverImpl` isn't a trackable type, but it implements `IWhatever`, and it would then register the whole implementation auto-magically and without your intervention. 

Interface read-write properties would be like tracking configurations:

```
// This would be a pseudo-code of what TrackerDog would do internally.
// It would implement a new non-generic registration API

// The object type yet to be trackable would be registered using the non-generic
// API 
var trackableTypeRegistration = Track.ThisType(trackableObject.GetType());

// Each property in the whole IWhatever interface would produce an auto-registration
// of its implemented counterpart in the given implementation type:
foreach(PropertyInfo property in typeof(IWhatever).GetProperties())
{
       trackableTypeRegistration.IncludeProperty(property);
}
```

-----

It seems like the 2nd option is the most powerful one and I would go this way, because we all want less configuration, don't we?

I'll try to implement this 2nd option next week and I'll notice you @ordinaryorange  and @FranckDepoortere  once you can give a try. I expect that it won't be hard to implement but you know, some small changes are very powerful!

**Comment by ordinaryorange** [comment]
Yes, option 2 is definitely the better option. Option 1 would not fit well with Inversion of Control frameworks. Option 2 is definitely worth the wait.

**Comment by FranckDepoortere** [comment]
I'm looking forward to testing the Lazy Registration.

**Comment by mfidemraizer** [comment]
Ok, I've a preview build assembly which already implements this feature. I've not tested yet very well, but this very basic test already works:

    [TestClass]
    public sealed class BaseTypeTrackingTest
    {
        public interface IWhatever
        {
            string Text { get; set; }
        }

        public class WhateverImpl : IWhatever
        {
            public virtual string Text { get; set; }
        }

        [ClassInitialize]
        public static void Init(TestContext context)
        {
            TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<IWhatever>());
        }

        [TestMethod]
        public void CanTrackInterfaceImplementations()
        {
           // WhateverImpl instance will be a proxy like if you would have configured it manually
            WhateverImpl whatever = new WhateverImpl().AsTrackable();
            whatever.Text = "hello world"; 
        }
    }

In the other hand, I understand that we just want to auto-configure properties from the given interface. 

It would be strange that you would track implementations of `IWhatever` and TrackerDog could also map properties that aren't part of the whole interface. Do you follow me? ;P

If you want to test this feature now, I've attached the following ZIP with the assembly:

[TrackerDog.zip](https://github.com/mfidemraizer/trackerdog/files/390761/TrackerDog.zip)

All previous tests are still passing so you can use it safely (unless the whole new feature, as being inmature yet, couldn't fully work...!).

Let me know, @ordinaryorange  @FranckDepoortere  if I'm in the right track!

**Comment by FranckDepoortere** [comment]
Hello @mfidemraizer ,

Thanks you for your improvement about Tracking Interfaces.

Nevertheless, I thing there a problem when we use autotracking on properties, and that a property has a Getter, and NO Setter.
I have hundreds of types of Objects to track and a cannot specify all properties to track.

Ex: 

```
interface IMyInterfaceId {
Guid Id  { get; }
string Name  { get; set; }
}
public class IdClass: IMyInterfaceId {
public virtual Guid Id  { get; }
public virtual string Name  { get; set; }
}

TrackerDogConfiguration.TrackTheseTypes(Track.ThisType<IMyInterfaceId>());

IMyInterfaceId whatever = new IMyInterfaceId().AsTrackable();
```

I've got a nullpointer line 69, class SimplePropertyInterceptionHook: p.SetMethod is NULL

 public bool ShouldInterceptMethod(Type type, MethodInfo methodInfo)
        { 
(.......)
                                    var accessorMethods = new[] { p.GetMethod.GetBaseDefinition(), p.SetMethod.GetBaseDefinition() };
        }

Franck

**Comment by mfidemraizer** [comment]
First of all, I've just released 2.0.0 on Nuget, now you can try it directly by upgrading your solution to this package version.

@FranckDepoortere About your concern, I would create a new issue to discuss this.

**Comment by mfidemraizer** [comment]
@FranckDepoortere Let's continue the discussion on #13
